homeassistant/helpers/selector.py

```diff
@@ -88,30 +88,37 @@ class EntitySelector(Selector):
             vol.Optional("domain"): str,
             # Device class of the entity
             vol.Optional("device_class"): str,
+            vol.Optional("multiple"): cv.boolean,
         }
     )
 
-    def __call__(self, data: Any) -> str:
+    def __call__(self, data: Any) -> str | list[str]:
         """Validate the passed selection."""
+        result: str | list[str]
+
+        def validate(e_or_u: str) -> None:
+            if "domain" in self.config:
+                if valid_entity_id(e_or_u):
+                    domain = split_entity_id(e_or_u)[0]
+                    if domain != self.config["domain"]:
+                        raise vol.Invalid(
+                            f"Entity {e_or_u} belongs to domain {domain}, "
+                            f"expected {self.config['domain']}"
+                        )
+
         try:
-            entity_id = cv.entity_id(data)
-            domain = split_entity_id(entity_id)[0]
+            result = cv.entity_id_or_uuid(cv.string(data))
+            validate(result)
         except vol.Invalid:
-            # Not a valid entity_id, maybe it's an entity entry id
-            return cv.entity_id_or_uuid(cv.string(data))
-        else:
-            if "domain" in self.config and domain != self.config["domain"]:
-                raise vol.Invalid(
-                    f"Entity {entity_id} belongs to domain {domain}, "
-                    f"expected {self.config['domain']}"
-                )
-
-            return entity_id
+            result = cv.entity_ids_or_uuids(data)
```

Comment:
When `multiple` is set to true, we should only validate it as multiple entities, and only validate as single if multiple is false.

Reply:
That is not always the case, selectors can also be used in places that can't handle lists.

Reply:
They selector config must specify multiple and the user must pick more than one to get a list.  So we should be OK.
The question is should we always return a list (even an empty list) if multiple is specified.

Reply:
Yes we should always return a list if multiple is set to true and a string if not. 

However, your current code is not checking if multiple is set, but just validates first as a single, then as multiple. We should just check the value of multiple.